File: src/main/java/ru/citeck/ecos/uiserv/domain/menu/api/records/MenuRecords.java
package ru.citeck.ecos.uiserv.domain.menu.api.records;

import ecos.com.fasterxml.jackson210.annotation.JsonProperty;
import ecos.com.fasterxml.jackson210.annotation.JsonValue;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import org.apache.commons.lang.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Component;
import ru.citeck.ecos.commons.data.ObjectData;
import ru.citeck.ecos.commons.json.Json;
import ru.citeck.ecos.records2.RecordMeta;
import ru.citeck.ecos.records2.RecordRef;
import ru.citeck.ecos.records2.graphql.meta.annotation.MetaAtt;
import ru.citeck.ecos.records2.graphql.meta.value.MetaField;
import ru.citeck.ecos.records2.request.delete.RecordsDelResult;
import ru.citeck.ecos.records2.request.delete.RecordsDeletion;
import ru.citeck.ecos.records2.request.mutation.RecordsMutResult;
import ru.citeck.ecos.records2.request.query.RecordsQuery;
import ru.citeck.ecos.records2.request.query.RecordsQueryResult;
import ru.citeck.ecos.records2.source.dao.local.LocalRecordsDao;
import ru.citeck.ecos.records2.source.dao.local.MutableRecordsLocalDao;
import ru.citeck.ecos.records2.source.dao.local.v2.LocalRecordsMetaDao;
import ru.citeck.ecos.records2.source.dao.local.v2.LocalRecordsQueryWithMetaDao;
import ru.citeck.ecos.uiserv.domain.menu.service.MenuService;
import ru.citeck.ecos.uiserv.domain.menu.dto.MenuDto;

import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Component
@RequiredArgsConstructor
public class MenuRecords extends LocalRecordsDao
    implements LocalRecordsMetaDao<MenuRecords.MenuRecord>,
               LocalRecordsQueryWithMetaDao<Object>,
               MutableRecordsLocalDao<MenuRecords.MenuRecord> {

    private static final String ID = "menu";
    private static final String AUTHORITIES_QUERY_LANG = "authorities";

    private final MenuService menuService;

    {
        setId(ID);
    }

    @NotNull
    @Override
    public List<MenuRecord> getValuesToMutate(@NotNull List<RecordRef> records) {
        return getLocalRecordsMeta(records, null);
    }

    @NotNull
    @Override
    public RecordsQueryResult<Object> queryLocalRecords(@NotNull RecordsQuery recordsQuery,
                                                        @NotNull MetaField metaField) {

        if (AUTHORITIES_QUERY_LANG.equals(recordsQuery.getLanguage())) {
            return new RecordsQueryResult<>(new ArrayList<>(menuService.getAllAuthoritiesWithMenu()));
        }

        if (!"predicate".equals(recordsQuery.getLanguage())
            && !"criteria".equals(recordsQuery.getLanguage())) {

            MenuQuery menuQuery = recordsQuery.getQuery(MenuQuery.class);
            if (menuQuery != null && StringUtils.isNotBlank(menuQuery.user)) {
                MenuDto menuDto = menuService.getMenuForUser(menuQuery.user);
                return RecordsQueryResult.of(new MenuRecord(menuDto));
            }
        }

        RecordsQueryResult<Object> result = new RecordsQueryResult<>();
        result.setRecords(menuService.getAllMenus()
            .stream()
            .map(MenuRecord::new)
            .collect(Collectors.toList()));

        return result;
    }

    @Override
    public RecordsMutResult save(List<MenuRecord> values) {

        RecordsMutResult result = new RecordsMutResult();
        values.forEach(dto -> {
            if (StringUtils.isBlank(dto.getId())) {
                throw new IllegalArgumentException("Parameter 'id' is mandatory for menu record");
            }

            if (dto.getId().equals("default-menu")) {
                dto.setId(UUID.randomUUID().toString());
            }

            MenuDto saved = menuService.save(dto);
            result.addRecord(new RecordMeta(saved.getId()));
        });

        return result;
    }

    @Override
    public RecordsDelResult delete(RecordsDeletion deletion) {
        RecordsDelResult result = new RecordsDelResult();
        deletion.getRecords().stream()
            .map(RecordRef::getId)
            .forEach(id -> {
                menuService.deleteByExtId(id);
                result.addRecord(new RecordMeta(id));
            });
        return result;
    }

    @Override
    public List<MenuRecord> getLocalRecordsMeta(List<RecordRef> records, MetaField metaField) {
        return records.stream()
            .map(RecordRef::getId)
            .map(id -> menuService.getMenu(id)
                .orElseGet(() -> new MenuDto(id)))
            .map(MenuRecord::new)
            .collect(Collectors.toList());
    }

    @Data
    public static class MenuQuery {
        private String user;
    }

    public static class MenuRecord extends MenuDto {

        public MenuRecord(MenuDto model) {
            super(model);
        }

        public MenuRecord() {
        }

        public String getModuleId() {
            return getId();
        }

        public void setModuleId(String value) {
            setId(value);
        }

        @MetaAtt(".disp")
        public String getDisplayName() {
            return getId();
        }

        @JsonProperty("_content")
        public void setContent(List<ObjectData> content) {

            String base64Content = content.get(0).get("url", "");
            base64Content = base64Content.replaceAll("^data:application/json;base64,", "");
            ObjectData data = Json.getMapper().read(Base64.getDecoder().decode(base64Content), ObjectData.class);

            Json.getMapper().applyData(this, data);
        }

        @JsonValue
        @com.fasterxml.jackson.annotation.JsonValue
        public MenuDto toJson() {
            return new MenuDto(this);
        }
    }
}

File: src/main/java/ru/citeck/ecos/uiserv/domain/menu/api/records/MenuRecords.kt
package ru.citeck.ecos.uiserv.domain.menu.api.records

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonValue
import jakarta.annotation.PostConstruct
import lombok.Data
import lombok.RequiredArgsConstructor
import org.apache.commons.lang3.StringUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import ru.citeck.ecos.commons.data.ObjectData
import ru.citeck.ecos.commons.json.Json.mapper
import ru.citeck.ecos.commons.json.YamlUtils.toNonDefaultString
import ru.citeck.ecos.events2.type.RecordEventsService
import ru.citeck.ecos.records2.RecordConstants
import ru.citeck.ecos.records2.predicate.PredicateService
import ru.citeck.ecos.records2.predicate.model.Predicate
import ru.citeck.ecos.records3.record.atts.schema.annotation.AttName
import ru.citeck.ecos.records3.record.atts.value.AttValue
import ru.citeck.ecos.records3.record.dao.AbstractRecordsDao
import ru.citeck.ecos.records3.record.dao.atts.RecordAttsDao
import ru.citeck.ecos.records3.record.dao.delete.DelStatus
import ru.citeck.ecos.records3.record.dao.delete.RecordDeleteDao
import ru.citeck.ecos.records3.record.dao.mutate.RecordMutateDtoDao
import ru.citeck.ecos.records3.record.dao.query.RecordsQueryDao
import ru.citeck.ecos.records3.record.dao.query.dto.query.RecordsQuery
import ru.citeck.ecos.records3.record.dao.query.dto.res.RecsQueryRes
import ru.citeck.ecos.uiserv.domain.i18n.service.MessageResolver
import ru.citeck.ecos.uiserv.domain.menu.api.records.MenuRecords.MenuMutRecord
import ru.citeck.ecos.uiserv.domain.menu.dto.MenuDto
import ru.citeck.ecos.uiserv.domain.menu.dto.SubMenuDef
import ru.citeck.ecos.uiserv.domain.menu.service.MenuService
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.stream.Collectors

@Component
@RequiredArgsConstructor
class MenuRecords(
    private val menuService: MenuService,
    private val messageResolver: MessageResolver
) : AbstractRecordsDao(),
    RecordAttsDao,
    RecordsQueryDao,
    RecordMutateDtoDao<MenuMutRecord>,
    RecordDeleteDao {

    companion object {
        const val ID = "menu"
        private const val AUTHORITIES_QUERY_LANG = "authorities"
    }

    private var recordEventsService: RecordEventsService? = null

    @PostConstruct
    fun init() {
        menuService.addOnChangeListener { before, after ->
            recordEventsService?.emitRecChanged(before, after, getId()) { MenuMutRecord(it) }
        }
    }

    override fun getId() = ID

    override fun getRecordAtts(recordId: String): MenuRecord {
        val menuDto = menuService.getMenu(recordId).orElseGet { MenuDto.EMPTY }
        return MenuRecord(menuDto)
    }

    override fun getRecToMutate(recordId: String): MenuMutRecord {
        return MenuMutRecord(getRecordAtts(recordId).model)
    }

    override fun queryRecords(recsQuery: RecordsQuery): RecsQueryRes<Any> {

        if (AUTHORITIES_QUERY_LANG == recsQuery.language) {
            return RecsQueryRes(ArrayList<Any>(menuService.allAuthoritiesWithMenu))
        }
        if (recsQuery.language == PredicateService.LANGUAGE_PREDICATE) {
            val predicate = recsQuery.getQuery(Predicate::class.java)
            val records = menuService.findAll(
                predicate,
                recsQuery.workspaces,
                recsQuery.page.maxItems,
                recsQuery.page.skipCount,
                recsQuery.sortBy
            ).map { MenuRecord(it) }
            val result = RecsQueryRes<Any>()
            result.setRecords(records)
            result.setTotalCount(menuService.getCount(predicate, recsQuery.workspaces))
            return result
        }
        if ("criteria" != recsQuery.language) {

            val menuQuery = recsQuery.getQueryOrNull(MenuQuery::class.java)

            if (menuQuery != null && StringUtils.isNotBlank(menuQuery.user)) {
                val menuDto = menuService.getMenuForCurrentUser(
                    menuQuery.version,
                    menuQuery.workspace ?: ""
                )
                return RecsQueryRes.of(MenuRecord(menuDto))
            }
        }

        val result = RecsQueryRes<Any>()
        result.setRecords(
            menuService.allMenus
                .stream()
                .map { model: MenuDto -> MenuRecord(model) }
                .collect(Collectors.toList())
        )
        return result
    }

    override fun saveMutatedRec(record: MenuMutRecord): String {
        require(!StringUtils.isBlank(record.id)) { "Parameter 'id' is mandatory for menu record" }
        if (MenuService.DEFAULT_MENUS.contains(record.id)) {
            record.id = UUID.randomUUID().toString()
        }
        val saved = menuService.save(record.build())
        return saved.id
    }

    override fun delete(recordId: String): DelStatus {
        menuService.deleteByExtId(recordId)
        return DelStatus.OK
    }

    @Autowired(required = false)
    fun setRecordEventsService(recordEventsService: RecordEventsService) {
        this.recordEventsService = recordEventsService
    }

    @Data
    class MenuQuery {
        val user: String? = null
        val version: Int? = null
        val workspace: String? = null
    }

    inner class MenuRecord(
        @AttName("...") val model: MenuDto
    ) {

        @AttName(RecordConstants.ATT_NOT_EXISTS)
        fun isNotExists(): Boolean {
            return model.id.isBlank()
        }

        fun getEcosType(): String {
            return "menu"
        }

        fun isDefaultMenu(): Boolean {
            return menuService.isDefaultMenu(model.id)
        }

        fun getDefaultMenuForJournal(): String {
            return messageResolver.getMessage(if (isDefaultMenu()) "label.yes" else "label.no")
        }

        fun getPermissions(): MenuPermissions {
            return MenuPermissions(!isDefaultMenu())
        }

        fun getModuleId(): String {
            return model.id
        }

        fun getDisplayName(): String {
            return model.id
        }

        fun getAuthoritiesForJournal(): String {
            val authorities = model.authorities
            return authorities.stream().filter { "GROUP_EVERYONE" != it }
                .collect(Collectors.joining(", "))
        }

        @JsonValue
        fun toJson(): Any {
            return model
        }

        fun getData(): ByteArray {
            return toNonDefaultString(toJson()).toByteArray(StandardCharsets.UTF_8)
        }
    }

    inner class MenuMutRecord(model: MenuDto?) : MenuDto.Builder(model ?: MenuDto.EMPTY) {

        override fun withSubMenu(subMenu: Map<String, SubMenuDef>?): MenuDto.Builder {
            val newSubMenu = HashMap(this.subMenu)
            subMenu?.forEach { (k, v) ->
                newSubMenu[k] = v
            }
            return super.withSubMenu(newSubMenu)
        }

        fun setModuleId(moduleId: String) {
            withId(moduleId)
        }

        @JsonProperty("_content")
        fun setContent(content: List<ObjectData>) {
            val dataUriContent = content[0].get("url", "")
            val data = mapper.read(dataUriContent, ObjectData::class.java)!!
            mapper.applyData(this, data)
        }
    }

    class MenuPermissions internal constructor(var editable: Boolean) : AttValue {
        override fun has(name: String): Boolean {
            return if ("Write" == name) {
                editable
            } else {
                true
            }
        }
    }
}
